File: typescript/packages/var-core/src/failure-anchor.ts
import { isCellMismatchError } from './cell-diff.js'
import { isDocStringMismatchError } from './doc-string-diff.js'
import type { Span } from './span.js'

// Where a failure POINTS in the .md: a mismatch anchors at its first failing
// span (the cell, the doc string fence body), anything else at the fallback —
// the step's match start. This rule is the single source of truth for failure
// locations: the executor's stack augmentation renders it per-runtime, and the
// conformance trace pins it as `failure.anchor`, so every language port must
// reproduce it byte-for-byte.
export function failureAnchor(error: unknown, fallback: Span): Span {
  if (isCellMismatchError(error)) return error.cells.find((c) => !c.ok)?.span ?? fallback
  if (isDocStringMismatchError(error)) return error.diff.span
  return fallback
}

File: python/packages/var-core/src/var_core/failure_anchor.py
"""failure_anchor.py — where a failure points in the .md source.

Port of failureAnchor from typescript/packages/var-core/src/failure-anchor.ts.
A mismatch anchors at its first failing span (the cell, the doc string fence
body), anything else at the fallback — the step's match start. This rule is
the single source of truth for failure locations: the executor's stack
augmentation renders it per-runtime, and the conformance trace pins it as
``failure.anchor``, so every language port must reproduce it byte-for-byte.
"""

from __future__ import annotations

from var_core.cell_diff import is_cell_mismatch_error
from var_core.doc_string_diff import is_doc_string_mismatch_error
from var_core.span import Span


def failure_anchor(error: object, fallback: Span) -> Span:
    if is_cell_mismatch_error(error):
        return next((c.span for c in error.cells if not c.ok), fallback)
    if is_doc_string_mismatch_error(error):
        return error.diff.span
    return fallback
